### PR TITLE
fixed load/users getting distributed to missing worker

### LINKED ISSUE #2008: Load being distributed to missing workers too Version 2.7.4.dev14
When running Locust in a distributed mode in Kubernetes, if the pod restarts then the worker status gets changed to missing. I can see that  spawning distributes the users equally to all workers, but in this case it distributes to missing workers as well which leads to reduced spwaning rate.
Kindly help!!
I have made some changes by adding Kubernetes support natively, have shared the details in slack channel. kindly suggest on how we can take this further.

### PR DESCRIPTION
**Pull request by radhakrishnaakamat**
fixed load/users getting distributed to the missing workers. 
I am new to open source commit and python. 
Please let me know if anything needs to be changed.
fixes #2008 #2010

### CODE REVIEW

**Comment by cyberw** [comment]
Can you add a unit test? You can override/shorten the timeout in the tests to reduce the time they take by using pytests mock feature (there should be some examples among the tests)

File: locust/test/test_dispatch.py [hunk 1]
```diff
@@ -3273,6 +3273,29 @@ class User5(User):
...
+
+        worker_nodes = [WorkerNode(str(i + 1)) for i in range(3)]
+        worker_nodes[1].heartbeat = HEARTBEAT_DEAD - 1
+
+        users_dispatcher = UsersDispatcher(worker_nodes=worker_nodes, user_classes=user_classes)
+
+        users_dispatcher.new_dispatch(target_user_count=9, spawn_rate=3)
+        users_dispatcher._wait_between_dispatch = 0
+        users_dispatcher.prepare_rebalance()
+        list(users_dispatcher)
```
Comment (cyberw) [inline locust/test/test_dispatch.py:3296]: what does this line do? :)

File: locust/test/test_dispatch.py [hunk 2]
```diff
@@ -3273,6 +3273,29 @@ class User5(User):
...
+        self.assertIs(len(users_dispatcher._worker_nodes), 2, "Expected worker 2 will be removed from worker nodes")
```
Comment (cyberw) [inline locust/test/test_dispatch.py:3297]: Can you also validate that the remaining workers get the right number of users?

**Comment by radhakrishnaakamat** [comment]
It asserts to Count of active workers in dispatch.. Here the missing worker
should be removed after heartdead interval.. Added comment to code and
added more sensible message in assert

File: locust/test/test_dispatch.py (hunk 2 above)
Comment (radhakrishnaakamat) [reply locust/test/test_dispatch.py]: added the test case

File: locust/test/test_dispatch.py (hunk 1 above)
Comment (radhakrishnaakamat) [reply locust/test/test_dispatch.py]: re-organized the code. it validates the workers available for redistribution

File: locust/dispatch.py [hunk 3]
```diff
@@ -16,6 +16,8 @@
 if TYPE_CHECKING:
     from locust.runners import WorkerNode
 
+HEARTBEAT_DEAD = -10
```
Comment (cyberw) [inline locust/dispatch.py:19]: import from runners instead?

File: locust/runners.py [hunk 4]
```diff
@@ -59,6 +59,7 @@
 CPU_MONITOR_INTERVAL = 5.0
 HEARTBEAT_INTERVAL = 1
 HEARTBEAT_LIVENESS = 3
+HEARTBEAT_DEAD = -10
```
Comment (cyberw) [inline locust/runners.py:62]: Feels weird that this is a negative number?

File: locust/dispatch.py (hunk 3 above)
Comment (radhakrishnaakamat) [reply locust/dispatch.py:19]: I tried that, but was getting error during initialization.
Will try the reverse tomorrow..

Could you help me with any resource on how to setup debugging in distributed mode.. That would be really helpful for me to contribute in a better way.. I am resorting to print statement for debugging(This is my first python project)

File: locust/runners.py (hunk 4 above)
Comment (radhakrishnaakamat) [reply locust/runners.py:62]: It symbolic to time with no response, as current heartbeat logic subtracts 1 for each missed heartbeat. Didn't want to introduce another variable to maintain skipped heart beat.

File: locust/dispatch.py (hunk 3 above)
Comment (cyberw) [reply locust/dispatch.py:19]: Sorry, I dont have much to contribute with debugging distributed. You know how to run an individual unit test, right?

Comment (radhakrishnaakamat) [reply locust/dispatch.py:19]: Yes, initially used to run tox.. later found out the way to run individually.

**Comment by mboutet** [comment]
Hi @radhakrishnaakamat, can you link this PR to the corresponding issue #2008?

**Comment by radhakrishnaakamat** [comment]
Fixes #<2010> #<2008>

**Comment by radhakrishnaakamat** [comment]
@mboutet thanks for the pointers in #2008, I was able to fix it in the right places in a better way. 
@cyberw could you kindly review it once again, I have updated the tests and tox worked fine too. Thanks!!

**Comment by cyberw** [comment]
👍👍

**Pull request merged**